crypto/store/store_lib.c

```diff
@@ -1095,3 +1095,62 @@ OSSL_STORE_CTX *OSSL_STORE_attach(BIO *bp, const char *scheme,
 
     return ctx;
 }
+
+const char *OSSL_file_stat(const char *uri, struct stat *st)
+{
+    const char *path = uri, *q;
+    struct stat local_st;
+
+    if (st == NULL)
+        st = &local_st;
+
+    /*
+     * First, unless the URI starts with "file://",
+     * try and see if the full URI can be taken as a local file path name.
+     */
+    if (!HAS_CASE_PREFIX(uri, "file://")) {
+        if (stat(path, st) == 0)
+            return uri;
+        ERR_raise_data(ERR_LIB_SYS, errno, "calling stat(%s)", path);
+    }
+
+    /* Do a second attempt only if the URI appears to start with the "file" scheme. */
+    if (!CHECK_AND_SKIP_CASE_PREFIX(path, "file:"))
+        return NULL;
+
+    /*
+     * Extract the alternative path to check.
+     * There's a special case if the URI also contains an authority,
+     * then the full URI shouldn't be used as a path anywhere.
+     */
+    q = path;
+    if (CHECK_AND_SKIP_CASE_PREFIX(q, "//")) {
+        if (CHECK_AND_SKIP_CASE_PREFIX(q, "localhost/")
+            || CHECK_AND_SKIP_CASE_PREFIX(q, "/")) {
+            /*
+             * In this case, we step back one char to ensure that the
+             * first slash is preserved, making the path always absolute
+             */
+            path = q - 1;
+        } else {
+            ERR_raise(ERR_LIB_OSSL_STORE, OSSL_STORE_R_URI_AUTHORITY_UNSUPPORTED);
+            return NULL;
+        }
+    }
+#ifdef _WIN32
+    /* Windows "file:" URIs with a drive letter start with a '/' */
+    if (path[0] == '/' && path[2] == ':' && path[3] == '/') {
+        char c = tolower((unsigned char)path[1]);
+
+        if (c >= 'a' && c <= 'z') {
+            /* Skip past the slash, making the path a normal Windows path */
+            path++;
+        }
+    }
+#endif
+
+    if (stat(path, st) == 0)
+        return path;
```

Comment:
BTW, it would probably be better to have this function return a bool rather than a local variable.  So replace all `return NULL` with `return 0`, and `return path` with `return 1`.

Reply:
This would

Reply:
Oh, and the the file store (both old and new implementation), the returned value is actually used.  So yeahok.